inspector/src/components/graph/canvasGraphComponent.tsx

```diff
@@ -14,8 +14,11 @@ export const CanvasGraphComponent: React.FC<ICanvasGraphComponentProps> = (props
         if (!canvasRef.current) {
             return;
         }
+
         // temporarily set empty array, will eventually be passed by props!
-        canvasServiceCallback(new CanvasGraphService(canvasRef.current, {datasets: []}));
+        const canvasGraphService = new CanvasGraphService(canvasRef.current, {datasets: []});
+        canvasServiceCallback(canvasGraphService);

```

Comment:
I am still troubled by this calling back to set it on the parent :-) and wonder if we could use a global state or send command to the component instead from the parent one ? I am afraid of ending up with the parent still holding to a destroyed service for instance ?

Reply:
If it has a destroyed service, would you want the ability to check if your service is destroyed like "isDestroyed"?

For global state, I am wondering what would happen if we want multiple CanvasComponents?  Could you elaborate more on this? Perhaps I'm misunderstanding the global state concept you're suggesting. 

Could you also elaborate a bit more about the other method you're suggesting?

My concerns for if we handle creation in the parent is how would we know when the canvas has been unmounted?

Reply:
To keep it at the simplest, a ref could work and you call functions on it ?

Reply:
Even if I totally understand it is not the "React Way" :-)

Reply:
@RaananW might have a better alternative

Reply:
> To keep it at the simplest, a ref could work and you call functions on it ?

So basically something like an action reducer which I send some Actions enum values? That could work.

I'll wait to see if @RaananW  also has any suggestions

Reply:
Seb and I talked about this offline, and I have made note of possible suggestions. Concerns make a lot more sense to me now. We will be keeping this callback for now, but will remove it the moment there is a bug that arises from it, or after the data layer is implemented (whichever comes first).

The idea is after the data layer is implemented we may be able to use observables to tell the child component "Hey I've updated the data, you can draw now with this new data".   Instead of having this circular dependency type approach. I may change this approach later on but it is the idea I am leaning towards.

Reply:
I always think of a service as a stateless singleton. It's simply a collection of functions with a common theme. In this case, this service is canvas-dependent, so it requires the canvas as a reference for each draw call (or, as you do it, sent in the constructor as a reference).
The "react way" would be to have those functions here, as they are a part of this component and technically have nothing to do with the parent. why does the parent call the draw function and not this component?

Reply:
Basically its because for now its easier to make changes in the parent without updating the state

Reply:
So to give a full summary for everyone else:

The reason I did this originally was because there was a sibling component which needed to share data between itself and the canvas.

After speaking with both Seb and Raanan the plan is to do the following:

Basically we will be having two observables: One for metadata, another for data (wrapped by some type of class inside of core)

So there will be a toggling/colourizing component that is a sibling to the canvas component which will be able to update the metadata, and will be a view on the metadata itself too.

Then there will be the canvas component which subscribes to changes to the data and the metadata, it will simply be a view for them.

I am marking this conversation as resolved, and to fix once the data class is finished!!